README.md
# QMXStdLib

## Synopsis

> QuantuMatriX Software Standard Library - V2.3.0
> Copyright (C) 2011-2019 QuantuMatriX Software, a QuantuMatriX Technologies Cooperative Partnership
>
> This library is meant to offer an array of general-purpose data structures and algorithms for a multitude of uses.  Each structure must be examined for its specific usefulness to be determined.  Please see the API documentation for more details, or visit the QMXStdLib website at 'https://git.qmx-software.com/open-source/qmxstdlib/'.
>
> 'QMXStdLib' is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

## Dependencies

> ### GNU Linux and Apple macOS
>
>> #### Compiler
>> * Clang >= 3.9 or GCC >= 7.0
>>
>> #### Documentation
>> * Doxygen >= 1.6
>>
>> #### Libraries
>> * Boost >= 1.69
>> * GTest >= 1.6 (optional)
>>
>> #### Tools
>> * CMake >= 3.0
>
> ### Microsoft Windows
>
>> #### Compiler
>> * MSYS2 + Clang >= 3.9 or MSYS2 + MinGW with GCC >= 7.0
>>
>> #### Documentation
>> * Doxygen >= 1.6
>>
>> #### Libraries
>> * Boost >= 1.69
>> * GTest >= 1.6 (optional)
>>
>> #### Tools
>> * CMake >= 3.0

## Building

> To build using CMake, use the following steps:
>
> * Navigate to the project's root directory, and create a new directory called 'build'.
> * Enter the 'build' directory and run the following command (replacing '[GENERATOR]' with the name of the desired CMake generator; note that only Clang and GCC-based compilers are currently supported):
>
>> cmake -G "[GENERATOR]" ..
>
> * After the project files are fully generated, build in the usual manner.  The following targets are supported:
>
>> * all (default; includes 'debug' and 'release')
>> * debug
>> * relwithdebinfo
>> * release
>> * minsizerel
>> * doc (builds Doxygen documentation)
>> * tests (includes 'tests_debug' and 'tests_release' )
>> * tests_debug
>> * tests_relwithdebinfo
>> * tests_release
>> * tests_minsizerel
>> * install (only targets actually built will be installed; unit tests are never installed)
>> * uninstall

## Installing and Uninstalling

> To install on any platform, use the 'install' target of the generated project files.  To uninstall on any platform, use the 'uninstall' target of the generated project files.  Note that on a GNU Linux platform, you will need to run 'ldconfig' after the installation process; otherwise, the system will not be able to locate the new libraries.

## CMake Package

> This project provides a CMake package which can be used with the 'find_package' command.  Once found, the following variables are set:
>
>> * QMXStdLib_VERSION (the current library version)
>> * QMXStdLib_LIB_DEBUG (the debug build of the library)
>> * QMXStdLib_LIB_RELEASE (the release build of the library)
>> * QMXStdLib_INCLUDE_PATH (the library headers path)
>> * QMXStdLib_LIB_PATH (the library path)

## Contacts and Support

> * Site: 'https://git.qmx-software.com/open-source/qmxstdlib/'
> * Forums: 'https://forums.qmx-software.com/qmxstdlib/' (Currently Unavailable)
> * Bug Tracker: 'https://git.qmx-software.com/open-source/qmxstdlib/issues'
> * Email: 'support@qmx-software.com'

## Event Messages

> | Event Index | Event Message                                                                             |
> |-------------|-------------------------------------------------------------------------------------------|
> | 00000000    | The object instance was already initialized!                                              |
> | 00000001    | The object instance was not initialized!                                                  |
> | 00000002    | The specified file was not a dynamic library!                                             |
> | 00000003    | The specified dynamic library could not be loaded!                                        |
> | 00000004    | The symbol could not be retrieved because the dynamic library was not loaded!             |
> | 00000005    | The symbol could not be retrieved!                                                        |
> | 00000006    | The dynamic library could not be unloaded because it was not loaded!                      |
> | 00000007    | The dynamic library could not be unloaded!                                                |
> | 00000008    | A malformed symbol was found in the path!                                                 |
> | 00000009    | A non-existent symbol was found in the path!                                              |
> | 0000000A    | The specified path could not be converted into a canonical path!                          |
> | 0000000B    | The specified path's symlink contents could not be retrieved!                             |
> | 0000000C    | The specified hardlink/symlink could not be created due to an invalid source target!      |
> | 0000000D    | The specified hardlink/symlink could not be created using the specified arguments!        |
> | 0000000E    | The copy operation could not be performed due to an invalid source target!                |
> | 0000000F    | The copy operation could not be completed!                                                |
> | 00000010    | The base was set to none, but the string format pointer was null!                         |
> | 00000011    | The specified config file could not be opened for reading!                                |
> | 00000012    | The specified config file could not be opened for writing!                                |
> | 00000013    | The current thread's ID was not set!                                                      |
> | 00000014    | The current thread's ID was an empty string!                                              |
> | 00000015    | The current thread was already added to the tracer!                                       |
> | 00000016    | The current thread was not added to the tracer!                                           |
> | 00000017    | The specified group could not be created because it already exists!                       |
> | 00000018    | The specified group could not be destroyed because it did not exist!                      |
> | 00000019    | The specified group could not be destroyed because it was not empty!                      |
> | 0000001A    | The thread barrier size could not be set because the specified group did not exist!       |
> | 0000001B    | The specified thread could not be destroyed because its group did not exist!              |
> | 0000001C    | The specified thread could not be destroyed because it did not exist!                     |
> | 0000001D    | The thread barrier could not be called because the specified group did not exist!         |
> | 0000001E    | The thread barrier could not be called because the specified group did not exist!         |
> | 0000001F    | The specified thread could not be joined because its group did not exist!                 |
> | 00000020    | The specified thread could not be joined because it did not exist!                        |
> | 00000021    | The threads could not be joined because the specified group did not exist!                |
> | 00000022    | The specified thread could not be interrupted because its group did not exist!            |
> | 00000023    | The specified thread could not be interrupted because it did not exist!                   |
> | 00000024    | The threads could not be interrupted because the specified group did not exist!           |
> | 00000025    | The specified thread could not be created because its group did not exist!                |
> | 00000026    | The specified thread could not be created because it already exists!                      |


CHANGELOG
Change Log
==========

Key
---
* = Info
~ = Change
- = Removal
+ = Addition



17FEB16 1.0.0 Robert M. Baker <malachy@qmxtech.com>
===================================================
* Initial release.



27FEB16 1.1.0 Robert M. Baker <malachy@qmxtech.com>
===================================================
~ Modified various classes to use configurable values.
~ Refactored code making use of the 'NonCopyable' mixin class to instead use C++11 method deletion.
~ Marked the 'NonCopyable' mixin class as deprecated in favor of C++11 method deletion.
~ Modified 'ProgramOptions' class to allow for config files with groups.
+ Added 'SaveConfigFile' method to 'ProgramOptions' class to facilitate writing of config options to file.
+ Added 'SetConfigOption' method to 'ProgramOptions' class to facilitate setting/adding of config options at runtime.
+ Added 'RemoveConfigOption' method to 'ProgramOptions' class to facilitate removal of config options at runtime.
+ Added the 'ThreadManager' class to ease creation and management of threads and thread groups.



27FEB16 1.1.1 Robert M. Baker <malachy@qmxtech.com>
===================================================
~ Refactored 'CreateGroup' and 'CreateThread' methods of 'ThreadManager' to be more optimal.



28FEB16 1.1.2 Robert M. Baker <malachy@qmxtech.com>
===================================================
~ Moved 'Create' method from 'Object' derived classes to 'Object' itself.
+ Added 'DoAllocate' boolean parameter to 'Create' method in 'Object' class to allow for deferment of object allocation.



29AUG19 2.0.0 Robert M. Baker <malachy@qmxtech.com>
===================================================
~ Refactored all code to conform to new QMX standards.
~ Modified 'clone' and 'cloneImp' methods from 'Object' class to accept a pointer to a derived class.
~ Uncommented code in 'toUTF8', 'toUTF16', and 'toUTF32' functions of 'String' namespace since GCC is now fully C++11 compliant.
~ Modified various class destructors to be non-virtual, as they are not intended for use as base classes.
- Removed 'cloneImp' method from 'DynamicLibrary' class, as cloning is not logical for this class.
- Removed 'operator=' and 'operatorAssignImp' methods from 'Object' class.
- Removed deprecated 'nonCopyable' class.
- Removed 'Plugin' mixin class, as it is being moved to QMXSAF.
- Removed 'PluginManager' class, as it is being moved to QMXSAF.
~ Converted 'UniqueRandom' class into a namespace with generate functions which accepts integral and floating point types.
+ Added standard type definitions for the 'Thread' class.
+ Added 'wrap' function to 'Numeric' namespace to facilitate wrapping of values within a closed interval.
+ Added 'getConfigOptions' method to 'ProgramOptions' class to facilitate retrieval of the underlying config map.
+ Added 'setConfigOptions' method to 'ProgramOptions' class to facilitate setting of the underlying config map.
+ Added 'getGroupIDs' method to 'ThreadManager' class to facilitate retrieval of all existing groups.
+ Added 'getThreadIDs' method to 'ThreadManager' class to facilitate retrieval of all existing threads within a group.
+ Added 'InputIterator', 'OutputIterator', 'ForwardIterator', 'BidirectionalIterator', and 'RandomAccessIterator' classes.



31AUG19 2.1.0 Robert M. Baker <malachy@qmxtech.com>
===================================================
~ Moved filesystem-related type definitions to 'FileSytem.hpp'.
~ Moved thread-related type definitions to 'ThreadManager.hpp' (except 'ThreadHash', 'SharedMutex, & 'SharedMutexPair').
~ Moved time-related type definitions to 'Timer.hpp'.



31AUG19 2.1.1 Robert M. Baker <malachy@qmxtech.com>
===================================================
~ Fixed various typos in the Doxygen documenation.



03SEP19 2.2.0 Robert M. Baker <malachy@qmxtech.com>
===================================================
~ Modifed all static member variables to be inline.
- Removed unnecessary source files 'ConsoleIO.cpp', 'StackTracer.cpp', and 'Utility.cpp'.



09SEP19 2.2.1 Robert M. Baker <malachy@qmxtech.com>
===================================================
~ Modified 'README.md' to reflect new dependency versions.



10SEP19 2.3.0 Robert M. Baker <malachy@qmxtech.com>
===================================================
~ Modified 'String::tokenize' function to accept an optional predicate argument.


include/Object.hpp
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Object.hpp
// Robert M. Baker | Created : 05JAN12 | Last Modified : 29AUG19 by Robert M. Baker
// Version : 2.3.0
// This is a header file for 'QMXStdLib'; it defines the interface for a generic base class.
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2011-2019 QuantuMatriX Software, a QuantuMatriX Technologies Cooperative Partnership
//
// This file is part of 'QMXStdLib'.
//
// 'QMXStdLib' is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free
// Software Foundation, either version 3 of the License, or (at your option) any later version.
//
// 'QMXStdLib' is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
// A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License along with 'QMXStdLib'.  If not, see <http://www.gnu.org/licenses/>.
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

/**
  * @file
  * @author  Robert M. Baker
  * @date    Created : 05JAN12
  * @date    Last Modified : 29AUG19 by Robert M. Baker
  * @version 2.3.0
  *
  * @brief This header file defines the interface for a generic base class.
  *
  * @section ObjectH0000 Description
  *
  * This header file defines the interface for a generic base class.
  *
  * @section ObjectH0001 License
  *
  * Copyright (C) 2011-2019 QuantuMatriX Software, a QuantuMatriX Technologies Cooperative Partnership
  *
  * This file is part of 'QMXStdLib'.
  *
  * 'QMXStdLib' is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free
  * Software Foundation, either version 3 of the License, or (at your option) any later version.
  *
  * 'QMXStdLib' is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  * A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
  *
  * You should have received a copy of the GNU Lesser General Public License along with 'QMXStdLib'.  If not, see <http://www.gnu.org/licenses/>.
  */

#ifndef __QMX_QMXSTDLIB_OBJECT_HPP_
#define __QMX_QMXSTDLIB_OBJECT_HPP_

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Header Files
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "Base.hpp"
#include "Mixins/Lockable.hpp"
#include "RAII/ScopedLock.hpp"
#include "RAII/ScopedStackTrace.hpp"

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Start of the 'QMXStdLib' Namespace
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

namespace QMXStdLib
{

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// The 'NoPublicProperties' Structure
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

/**
  * @brief This structure is for use in cases where a class derived from 'Object' has no public properties.
  */

struct NoPublicProperties
{
};

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// The 'Object' Class
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

/**
  * @class Object Object.hpp "include/Object.hpp"
  *
  * @brief This class defines base algorithms for all classes.
  *
  * The purpose of this generic base-class is to encapsulate functionality useful to all sub-classes.
  *
  * Platform Independent     : Yes<br>
  * Architecture Independent : Yes<br>
  * Thread-Safe              : Yes
  */

template< typename DType, typename PPType = NoPublicProperties > class Object : public Lockable
{
public:

	// Public Type Definitions

		STANDARD_TYPEDEFS( DType )
		typedef PPType PublicPropertiesType;

	// Destructor

		/**
		  * @brief This is the destructor.
		  */

		virtual ~Object()
		{
			// Do nothing.
		}

	// Public Methods

		/**
		  * @brief This method determines if this object is initialized.
		  *
		  * @return
		  * 	A boolean value of 'true' if this object is initialized, and 'false' otherwise.
		  */

		bool isObjectInitialized() const
		{
			// Report wether or not this object is initialized to calling routine.

				return initialized;
		}

		/**
		  * @brief This method performs necessary initialization.
		  *
		  * @exception QMXException
		  * 	If the object instance was already initialized.
		  */

		void allocate()
		{
			// Obtain locks.

				SCOPED_WRITE_LOCK;

			// Create scoped stack traces.

				SCOPED_STACK_TRACE( "Object::allocate", 0000 );

			// Perform necessary initialization.

				QMX_ASSERT( !initialized, "QMXStdLib", "Object::Allocate", "00000000", "" );
				allocateImp();
				initialized = true;
		}

		/**
		  * @brief This method performs necessary cleanup.
		  *
		  * @exception QMXException
		  * 	If the object instance was not initialized.
		  */

		void deallocate()
		{
			// Obtain locks.

				SCOPED_WRITE_LOCK;

			// Create scoped stack traces.

				SCOPED_STACK_TRACE( "Object::deallocate", 0000 );

			// Perform necessary cleanup.

				QMX_ASSERT( initialized, "QMXStdLib", "Object::deallocate", "00000001", "" );
				deallocateImp();
				initialized = false;
		}

		/**
		  * @brief This is the public properties accessor for this class.
		  *
		  * @param target
		  * 	This is a 'PPType' structure which will be set using the public properties of 'this'.
		  */

		void get( PPType& target ) const
		{
			// Obtain locks.

				SCOPED_READ_LOCK;

			// Set specified 'PPType' structure using the public properties of 'this'.

				target = properties;
		}

		/**
		  * @brief This is the public properties mutator for this class.
		  *
		  * @param source
		  * 	This is a 'PPType' structure which will be used to set the public properties of 'this'.
		  */

		void set( const PPType& source )
		{
			// Obtain locks.

				SCOPED_WRITE_LOCK;

			// Set public properties of 'this' using the specified 'PPType' structure.

				properties = source;
		}

		/**
		  * @brief This method creates an instance of this class.
		  *
		  * @param doAllocate
		  * 	This is a boolean value which determines if 'allocate' is called on the created instance.
		  *
		  * @return
		  * 	A pointer to the newly created instance.
		  *
		  * @exception QMXException
		  * 	If the an error occurs during allocation.
		  */

		static InstancePtr create( const bool doAllocate = true )
		{
			// Create scoped stack traces.

				SCOPED_STACK_TRACE( "Object::create", 0000 );

			// Create local variables.

				InstancePtr result( ( new DType() ) );

			// Initialize new instance, if 'doAllocate' is 'true'.

				if( doAllocate )
					result->allocate();

			// Return result to calling routine.

				return result;
		}

		/**
		  * @brief This method assigns the data of 'this' to the specified object.
		  *
		  * If the specified object pointer is the same as 'this', this method will have no effect.
		  *
		  * @param target
		  * 	This is the object pointer to use when setting.
		  */

		void clone( InstancePtr& target ) const
		{
			// Perform abort check.

				if( target.get() == this )
					return;

			// Obtain locks.

				SCOPED_READ_LOCK;
				SCOPED_WRITE_LOCK_X( target, 0000 );

			// Assign data of 'this' to specified object.

				target->initialized = initialized;
				target->properties = properties;
				cloneImp( target );
		}

protected:

	// Protected Fields

		/**
		  * @brief This is the flag used to determine initialization state.
		  */

		bool initialized;

		/**
		  * @brief These are the public properties of this class.
		  */

		PPType properties;

	// Protected Methods

		/**
		  * @brief This is the default constructor, which is made protected to prevent direct instantiation.
		  */

		Object()
		{
			// Initialize fields.

				initialized = false;
		}

private:

	// Private Methods

		/**
		  * @brief This is the overridable implementation for the 'allocate' method.
		  */

		virtual void allocateImp()
		{
			// Do nothing.
		}

		/**
		  * @brief This is the overridable implementation for the 'deallocate' method.
		  */

		virtual void deallocateImp()
		{
			// Do nothing.
		}

		/**
		  * @brief This is the overridable implementation for the 'clone' method.
		  *
		  * @param target
		  * 	This is the object pointer to use when setting.
		  */

		virtual void cloneImp( InstancePtr& target ) const
		{
			// Do nothing.
		}
};

} // 'QMXStdLib' Namespace

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// End of the 'QMXStdLib' Namespace
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#endif // __QMX_QMXSTDLIB_OBJECT_HPP_

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// End of 'Object.hpp'
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
